pyzim/pyzim.pyx

```diff
@@ -1,338 +1,183 @@
+    bool bool_cy_call_fct(void *ptr, string method, int *error):
+        """Lookup and execute a pure virtual method on ZimArticle returning a bool"""
 
-    @property
-    def filename(self):
-        return self._filename
-
-    def get_article(self, url):
-        # Read to a zim::Article
-        cdef zim.Article art = self.c_file.getArticleByUrl(url.encode('UTF-8'))
-        if not art.good():
-            raise RuntimeError("Article not found for url")
-
-        article = ZimArticle.from_read_article(art)
-        return article
-
-    def get_metadata(self):
-        metadata = dict()
-        for key in self._metadata_keys:
-            try:
-                meta_art = self.get_article(f"M/{key}")
-            except:
-                pass
-            else:
-                metadata[key] = meta_art.content
-        return metadata
-
-    def get_article_by_id(self, id):
-        # Read to a zim::Article
-        cdef zim.Article art = self.c_file.getArticle(<int> id)
-        if not art.good():
-            raise RuntimeError("Article not found for url")
-
-        article = ZimArticle.from_read_article(art)
-        return article
-
-    def get_redirect_article(self, ZimArticle article):
-        cdef zim.Article art = self.c_file.getArticleByUrl(article.redirect_longurl.encode('UTF-8'))
-
-        if article.is_redirect:
-            if not art.good():
-                raise RuntimeError("Article not found for redirect url")
-            article = ZimArticle.from_read_article(art)
-            return article
+        cdef ZimArticle art = <ZimArticle>(ptr)
+        try:
+            func = getattr(art, method.decode('UTF-8'))
+        except AttributeError:
+            error[0] = 1
+            raise 
         else:
-            raise RuntimeError("Article is not a redirect article")
+            error[0] = 0
+            return func()
+
+    uint64_t int_cy_call_fct(void *ptr, string method, int *error):
+        """Lookup and execute a pure virtual method on ZimArticle returning an int"""
 
-    def get_main_page_url(self):
-        cdef zim.Fileheader header = self.c_file.getFileheader()
-        cdef zim.Article article
-        if header.hasMainPage():
-            article = self.c_file.getArticle(header.getMainPage())
-            return article.getLongUrl().decode("UTF-8", "strict");
+        cdef ZimArticle art = <ZimArticle>(ptr)
+        try:
+            func = getattr(art, method.decode('UTF-8'))
+        except AttributeError:
+            error[0] = 1
+            raise 
+        else:
+            error[0] = 0
+            return <uint64_t> func()
 
-        # TODO Ask about the old format, check libzim for tests
-        # Handle old zim where header has no mainPage.
-        # (We need to get first article in the zim)
-        article = self.c_file.getArticle(<int> 0)
-        if article.good():
-            return article.getLongUrl().decode("UTF-8", "strict")
+class ZimTestArticle(ZimArticle):
```

Comment:
A `ZimTestArticle` should not be in the library definition.
At best in a example implementation (but in a different directory that the package src)

Reply:
Agree. It's currently there on the WIP.